### PR TITLE
Remove `rz_str_newf` from RzAsm Plugins

### PR DESCRIPTION
**Pull request by HN026**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Partially addresses #4127 
...

### CODE REVIEW

File: librz/asm/p/asm_arm_cs.c [hunk 1]
```diff
@@ -183,24 +183,22 @@ static int disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
...
 		}
-		char *buf_asm = rz_str_newf("%s%s%s",
-			insn->mnemonic,
-			insn->op_str[0] ? " " : "",
-			insn->op_str);
-		if (buf_asm) {
+		if (insn) {
 			if (!disp_hash) {
-				rz_str_replace_char(buf_asm, '#', 0);
+				rz_str_replace_char(insn->op_str, '#', 0);
```
Comment (wargio) [inline librz/asm/p/asm_arm_cs.c]: no you should not do this. you can pass the pointer of `rz_asm_op_get_asm`

Comment (wargio) [reply librz/asm/p/asm_arm_cs.c]: but `rz_asm_op_setf_asm` needs to be set before this invocation.

File: librz/asm/p/asm_chip8.c [hunk 2]
```diff
@@ -11,82 +11,81 @@ static int chip8_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *b, int l) {
...
+		case 0x18: rz_asm_op_setf_asm(op, "ld st, v%1x", x); break;
+		case 0x1E: rz_asm_op_setf_asm(op, "add i, v%1x", x); break;
+		case 0x29: rz_asm_op_setf_asm(op, "ld f, v%1x", x); break;
+		case 0x33: rz_asm_op_setf_asm(op, "ld b, v%1x", x); break;
+		case 0x55: rz_asm_op_setf_asm(op, "ld [i], v%1x", x); break;
+		case 0x65: rz_asm_op_setf_asm(op, "ld v%1x, [i]", x); break;
+		case 0x30: rz_asm_op_setf_asm(op, "ld hf, v%1x", x); break;
+		case 0x75: rz_asm_op_setf_asm(op, "ld r, v%1x", x); break;
+		case 0x85: rz_asm_op_setf_asm(op, "ld v%1x, r", x); break;
 		}
```
Comment (wargio) [inline librz/asm/p/asm_chip8.c:81]: add a default case in this switch which sets `invalid` like below.

File: librz/asm/p/asm_chip8.c [hunk 3]
```diff
@@ -11,82 +11,81 @@ static int chip8_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *b, int l) {
...
+	default: rz_asm_op_set_asm(op, "invalid"); break;
 	}
-	if (!buf_asm) {
-		rz_strbuf_set(&op->buf_asm, "invalid");
-	} else {
-		rz_strbuf_set(&op->buf_asm, buf_asm);
-		free(buf_asm);
+
+	if (!*rz_strbuf_get(&op->buf_asm)) {
+		rz_asm_op_set_asm(op, "invalid");
```
Comment (wargio) [inline librz/asm/p/asm_chip8.c]: i don't think this will be needed if all the cases have a `invalid` default case.

Comment (wargio) [reply librz/asm/p/asm_chip8.c]: or we could set `invalid` at the beginning and allow to change it again when it gets overwritten by the next set_asm

File: librz/asm/p/asm_riscv_cs.c [hunk 4]
```diff
@@ -43,12 +43,11 @@ static int riscv_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
 		goto fin;
 	}
 	op->size = insn->size;
-	char *str = rz_str_newf("%s%s%s", insn->mnemonic, insn->op_str[0] ? " " : "", insn->op_str);
+	rz_asm_op_setf_asm(op, "%s%s%s", insn->mnemonic, insn->op_str[0] ? " " : "", insn->op_str);
+	char *str = rz_strbuf_get(&op->buf_asm);
```
Comment (wargio) [inline librz/asm/p/asm_riscv_cs.c]:
```suggestion
	char *str = rz_asm_op_get_asm(op);
```

File: librz/asm/p/asm_sparc_cs.c [hunk 5]
```diff
@@ -52,14 +52,13 @@ static int sparc_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
 		goto fin;
 	}
 	op->size = insn->size;
-	char *buf_asm = rz_str_newf("%s%s%s",
+	rz_asm_op_setf_asm(op, "%s%s%s",
 		insn->mnemonic, insn->op_str[0] ? " " : "",
 		insn->op_str);
+	char *buf_asm = rz_strbuf_get(&op->buf_asm);
```
Comment (wargio) [inline librz/asm/p/asm_sparc_cs.c]:
```suggestion
	char *buf_asm = rz_asm_op_get_asm(op);
```

**Review by wargio — CHANGES_REQUESTED** [review]
very good

File: librz/asm/p/asm_mips_cs.c [hunk 6]
```diff
@@ -60,12 +60,11 @@ static int mips_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
 		goto fin;
 	}
 	op->size = insn->size;
-	char *str = rz_str_newf("%s%s%s", insn->mnemonic, insn->op_str[0] ? " " : "", insn->op_str);
+	rz_asm_op_setf_asm(op, "%s%s%s", insn->mnemonic, insn->op_str[0] ? " " : "", insn->op_str);
+	char *str = rz_strbuf_get(&op->buf_asm);
 	if (str) {
 		rz_str_replace_char(str, '$', 0);
 		// remove the '$'<registername> in the string
```
Comment (XVilka) [inline librz/asm/p/asm_mips_cs.c:66]: Move this comment one line up, before the actual function that does it.

File: librz/asm/p/asm_riscv_cs.c [hunk 7]
```diff
@@ -43,12 +43,11 @@ static int riscv_disassemble(RzAsm *a, RzAsmOp *op, const ut8 *buf, int len) {
...
 	if (str) {
 		rz_str_replace_char(str, '$', 0);
 		// remove the '$'<registername> in the string
```
Comment (XVilka) [inline librz/asm/p/asm_riscv_cs.c:49]: Move this comment one line up, before the actual function that does it.

**Review by XVilka — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**